File: android/app/src/main/kotlin/org/fgsec/halogen/storage/Outbox.kt
package org.fgsec.halogen.storage

import java.util.UUID
import org.fgsec.halogen.components.ToastCenter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.yield
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.halogen_mobile.SyncQueue
import uniffi.halogen_mobile.SyncDrainReport
import uniffi.halogen_mobile.openSyncQueue
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.fgsec.halogen.core.DeviceLog
import org.fgsec.halogen.wire.OrderDirection
import org.fgsec.halogen.wire.PlaylistReorderField
import org.fgsec.halogen.wire.PodcastConfigUpdateData

/// A queued offline mutation — the native analog of the web outbox op
/// vocabulary; ops serialize (kotlinx polymorphic, SerialName = the iOS case
/// name) and persist across relaunches.
@Serializable
data class OutboxOp(val id: String, val kind: Kind) {
    constructor(kind: Kind) : this(UUID.randomUUID().toString(), kind)

    @Serializable
    sealed interface Kind {
        // Playback
        @Serializable
        @SerialName("setCursor")
        data class SetCursor(val episodeId: Int, val cursor: ULong) : Kind

        @Serializable
        @SerialName("setPlayed")
        data class SetPlayed(val episodeId: Int, val played: Boolean) : Kind

        // Playlist membership
        @Serializable
        @SerialName("addToPlaylist")
        data class AddToPlaylist(
            val playlistId: Int, val episodeId: Int, val position: Int? = null
        ) : Kind

        @Serializable
        @SerialName("removeFromPlaylist")
        data class RemoveFromPlaylist(val playlistId: Int, val episodeId: Int) : Kind

        @Serializable
        @SerialName("moveInPlaylist")
        data class MoveInPlaylist(val playlistId: Int, val episodeId: Int, val to: Int) : Kind

        // Playlist structure. `updatePlaylist` is queued only when offline —
        // online edits go direct so forms can show server errors (web rule).
        // Later optional fields default to null for ops persisted before they
        // existed (same back-compat rule as the web's serde defaults).
        @Serializable
        @SerialName("reorderPlaylist")
        data class ReorderPlaylist(
            val playlistId: Int, val field: PlaylistReorderField, val direction: OrderDirection
        ) : Kind

        @Serializable
        @SerialName("updatePlaylist")
        data class UpdatePlaylist(
            val playlistId: Int,
            val name: String? = null,
            val isDefault: Boolean? = null,
            val description: String? = null,
            val deleteServerFile: Boolean? = null,
            val deleteClientFile: Boolean? = null,
        ) : Kind

        @Serializable
        @SerialName("movePlaylist")
        data class MovePlaylist(val playlistId: Int, val to: Int) : Kind

        // Library
        @Serializable
        @SerialName("subscribe")
        data class Subscribe(
            val feedUrl: String, val title: String? = null, val description: String? = null
        ) : Kind

        @Serializable
        @SerialName("unsubscribe")
        data class Unsubscribe(val podcastId: Int) : Kind

        // Server downloads
        @Serializable
        @SerialName("triggerDownload")
        data class TriggerDownload(val episodeId: Int) : Kind

        @Serializable
        @SerialName("removeServerDownload")
        data class RemoveServerDownload(val episodeId: Int) : Kind

        // Podcast config (update/remove target existing ids — safe to drain
        // later; create stays online-only, it needs a real id back).
        @Serializable
        @SerialName("updatePodcastConfig")
        data class UpdatePodcastConfig(val configId: Int, val data: PodcastConfigUpdateData) : Kind

        @Serializable
        @SerialName("removePodcastConfig")
        data class RemovePodcastConfig(val podcastId: Int) : Kind

        // `addToStart` defaults to null for pre-field persisted ops.
        @Serializable
        @SerialName("setAutoPlaylists")
        data class SetAutoPlaylists(
            val podcastId: Int, val playlistIds: List<Int>, val addToStart: Boolean? = null
        ) : Kind
    }
}

/** Native UI adapter over the shared Rust journal and retry engine. */
class Outbox private constructor(
    private val store: LocalStore,
    private val queue: SyncQueue,
    private val perform: suspend (SyncQueue) -> SyncDrainReport,
    private val pull: suspend (SyncQueue) -> String?,
    private val onDeadLetter: suspend (OutboxOp, Throwable) -> Unit,
    initial: List<OutboxOp>,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val gate = JournalGate.forPath(store.syncDatabasePath)
    private val ops = initial.toMutableList()
    private var draining = false
    @Volatile var suspended = false
        private set

    suspend fun pendingCount(): Int = mutex.withLock { ops.size }

    /// Whether any queued op still targets `playlistId`'s membership/order —
    /// refresh guards keep the local list authoritative until these drain.
    suspend fun hasPendingOps(playlistId: Int): Boolean = mutex.withLock {
        ops.any {
            when (val kind = it.kind) {
                is OutboxOp.Kind.AddToPlaylist -> kind.playlistId == playlistId
                is OutboxOp.Kind.RemoveFromPlaylist -> kind.playlistId == playlistId
                is OutboxOp.Kind.MoveInPlaylist -> kind.playlistId == playlistId
                is OutboxOp.Kind.ReorderPlaylist -> kind.playlistId == playlistId
                else -> false
            }
        }
    }

    /// Whether an Unsubscribe for `podcastId` is still queued — the podcasts
    /// tombstone set keeps its entries until their op drains.
    suspend fun hasPendingUnsubscribe(podcastId: Int): Boolean = mutex.withLock {
        ops.any { (it.kind as? OutboxOp.Kind.Unsubscribe)?.podcastId == podcastId }
    }

    /// Whether any queued op still reorders the playlists themselves — the
    /// playlists-list refresh keeps the local order until the moves drain.
    suspend fun hasPendingPlaylistMoves(): Boolean = mutex.withLock {
        ops.any { it.kind is OutboxOp.Kind.MovePlaylist }
    }

    suspend fun clearAll() {
        gate.withLock { mutex.withLock {
            queue.clearAll()
            ops.clear()
            store.saveDurably(ops.toList(), KEY)
        } }
    }

    suspend fun enqueue(kind: OutboxOp.Kind): Boolean = enqueueBatch(listOf(kind))

    suspend fun enqueueBatch(kinds: List<OutboxOp.Kind>): Boolean {
        var journaled = false
        try {
            gate.withLock { mutex.withLock {
                val operations = kinds.map(::OutboxOp)
                queue.importOperations(operations.map { it.sharedOperation() })
                journaled = true
                for (operation in operations) store.project(operation)
                val pending = queue.pendingIds().toSet()
                ops.removeAll { it.id !in pending }
                ops.addAll(operations.filter { it.id in pending })
                // Rebuilt from canonical pending entries if this derived mirror write fails.
                store.save(ops.toList(), KEY)
            } }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            DeviceLog.error("sync persistence failed: $error")
            ToastCenter.error(if (journaled) "The change is queued, but its local cache could not update. Reopen the app to recover it." else "Couldn't save this change. Please try again.")
            return false
        }
        scope.launch { yield(); drain() }
        return true
    }

    suspend fun pendingOperations(): List<OutboxOp> = mutex.withLock { ops.toList() }

    suspend fun setSuspended(value: Boolean) { mutex.withLock { suspended = value } }

    suspend fun drain() {
        mutex.withLock { if (suspended || draining) return; draining = true }
        val failures = mutableListOf<Pair<OutboxOp, Throwable>>()
        try {
            gate.withLock {
                if (suspended) return@withLock
                val submitted = mutex.withLock { ops.toList() }
                queue.importOperations(submitted.map { it.sharedOperation() })
                val report = perform(queue)
                projectAndCheckpoint(store, queue)
                if (!suspended && !report.authPaused) pull(queue)?.let { store.projectSnapshot(it) }
                val rejected = queue.quarantinedIds().toSet()
                val pending = queue.pendingIds().toSet()
                mutex.withLock {
                    for (op in ops.filter { it.id in rejected }) failures.add(op to IllegalStateException(report.lastError ?: "Change rejected by the server"))
                    ops.removeAll { it.id !in pending }
                    store.saveDurably(ops.toList(), KEY)
                }
            }
        } catch (error: CancellationException) { throw error }
        catch (error: Exception) { DeviceLog.warn("sync paused: $error") }
        finally { mutex.withLock { draining = false } }
        for ((operation, error) in failures) onDeadLetter(operation, error)
    }

    companion object {
        private const val KEY = "outbox"
        private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

        private suspend fun projectAndCheckpoint(store: LocalStore, queue: SyncQueue) {
            val delivered = queue.deliveredIds().toSet()
            val confirmed = mutableListOf<String>()
            for (operation in restoreSharedQueue(queue.pending())) {
                if (store.project(operation) && operation.id in delivered) confirmed.add(operation.id)
            }
            if (confirmed.isNotEmpty()) {
                queue.confirmCached(confirmed)
                store.forgetJournalMarkers(confirmed.toSet())
            }
        }

        suspend fun create(
            store: LocalStore,
            perform: suspend (SyncQueue) -> SyncDrainReport,
            pull: suspend (SyncQueue) -> String?,
            onDeadLetter: suspend (OutboxOp, Throwable) -> Unit = { _, _ -> },
        ): Outbox {
            val failures = mutableListOf<OutboxOp>()
            val outbox = JournalGate.forPath(store.syncDatabasePath).withLock {
                val original = store.prepareOutboxMigration()
                val ops = original?.let { json.decodeFromString<List<OutboxOp>>(it) } ?: emptyList()
                val queue = withContext(Dispatchers.IO) { openSyncQueue(store.syncDatabasePath) }
                queue.importOperations(ops.map { it.sharedOperation() })
                queue.cachedSnapshot()?.let { store.projectSnapshot(it) }
                val rejected = queue.quarantinedIds().toSet()
                failures.addAll(ops.filter { it.id in rejected })
                projectAndCheckpoint(store, queue)
                val active = queue.pendingIds().toSet()
                val remaining = restoreSharedQueue(queue.pending()).filter { it.id in active }
                store.saveDurably(remaining, KEY)
                Outbox(store, queue, perform, pull, onDeadLetter, remaining)
            }
            for (op in failures) onDeadLetter(op, IllegalStateException("Change rejected by the server"))
            return outbox
        }
    }
}

File: ios/Halogen/Storage/Outbox.swift
import Foundation

/// A queued offline mutation — Swift analog of the web's `OutboxOp` vocabulary,
/// Codable and persisted across relaunches. Op-for-op web mapping (single-id
/// download ops where the web batches; the endpoints are idempotent).
struct OutboxOp: Codable, Identifiable {
    enum Kind: Codable {
        // Playback
        case setCursor(episodeId: Int32, cursor: UInt64)
        case setPlayed(episodeId: Int32, played: Bool)
        // Playlist membership
        case addToPlaylist(playlistId: Int32, episodeId: Int32, position: Int32?)
        case removeFromPlaylist(playlistId: Int32, episodeId: Int32)
        case moveInPlaylist(playlistId: Int32, episodeId: Int32, to: Int32)
        // Playlist structure. `updatePlaylist` is queued only when offline —
        // online edits go direct so forms can show server errors (web rule).
        // Its later optional fields decode as nil for ops persisted before
        // they existed (same back-compat rule as the web's serde defaults).
        case reorderPlaylist(
            playlistId: Int32, field: PlaylistReorderField, direction: OrderDirection)
        case updatePlaylist(
            playlistId: Int32, name: String?, isDefault: Bool?,
            description: String?, deleteServerFile: Bool?, deleteClientFile: Bool?)
        case movePlaylist(playlistId: Int32, to: Int32)
        // Library
        case subscribe(feedUrl: String, title: String?, description: String?)
        case unsubscribe(podcastId: Int32)
        // Server downloads
        case triggerDownload(episodeId: Int32)
        case removeServerDownload(episodeId: Int32)
        // Podcast config (update/remove target existing ids — safe to drain
        // later; create stays online-only, it needs a real id back).
        case updatePodcastConfig(configId: Int32, data: PodcastConfigUpdateData)
        case removePodcastConfig(podcastId: Int32)
        // `addToStart` decodes as nil for pre-field persisted ops (web:
        // `#[serde(default)]` on SetPodcastAutoPlaylists.add_to_start).
        case setAutoPlaylists(podcastId: Int32, playlistIds: [Int32], addToStart: Bool?)
    }

    let id: UUID
    let kind: Kind

    init(_ kind: Kind) {
        self.id = UUID()
        self.kind = kind
    }
}

/// Native UI adapter over the shared Rust journal and retry engine.
actor Outbox {
    private static let key = "outbox"
    private let store: LocalStore
    private let queue: SyncQueue
    private let gate: JournalGate
    private let synchronize: (SyncQueue, String?) async throws -> String?
    private let perform: (SyncQueue) async throws -> SyncDrainReport
    private let onSnapshot: () async -> Void
    private let onDeadLetter: (OutboxOp, Error) async -> Void
    private var ops: [OutboxOp]
    private var draining = false
    private(set) var suspended = false

    init(
        store: LocalStore,
        perform: @escaping (SyncQueue) async throws -> SyncDrainReport,
        synchronize: @escaping (SyncQueue, String?) async throws -> String? = { _, _ in nil },
        onSnapshot: @escaping () async -> Void = {},
        onDeadLetter: @escaping (OutboxOp, Error) async -> Void = { _, _ in }
    ) async throws {
        self.store = store
        self.perform = perform
        self.synchronize = synchronize
        self.onSnapshot = onSnapshot
        self.onDeadLetter = onDeadLetter
        let path = await store.syncDatabasePath
        let gate = await JournalGate.forPath(path)
        self.gate = gate
        await gate.acquire()
        defer { Task { await gate.release() } }
        let original = try await store.prepareOutboxMigration()
        self.ops = try original.map { try WireJSON.decoder.decode([OutboxOp].self, from: $0) } ?? []
        self.queue = try openSyncQueue(dbPath: path)
        try await queue.importOperations(operations: ops.map { try $0.sharedOperation() })
        try await store.projectSnapshot(queue.cachedSnapshot())
        let rejected = Set(try await queue.quarantinedIds())
        for op in ops where rejected.contains(op.id.uuidString) {
            await onDeadLetter(
                op,
                NSError(
                    domain: "HalogenSync", code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Change rejected by the server"]))
        }
        let recovery = try OutboxOp.restoreSharedQueue(await queue.pending())
        try await store.projectJournal(recovery)
        let projected = Set(recovery.map { $0.id.uuidString })
        let delivered = try await queue.deliveredIds().filter { projected.contains($0) }
        try await queue.confirmCached(sourceIds: delivered)
        try await store.forgetJournalMarkers(Set(delivered))
        let pending = Set(try await queue.pendingIds())
        ops = recovery.filter { pending.contains($0.id.uuidString) }
        try await store.saveDurably(ops, key: Self.key)
    }

    var pendingCount: Int { ops.count }

    /// Whether any queued op still targets `playlistId`'s membership/order —
    /// refresh guards keep the local list authoritative until these drain
    /// (the web's `cache_playlists` membership-preservation invariant).
    func hasPendingOps(playlistId: Int32) -> Bool {
        ops.contains {
            switch $0.kind {
            case .addToPlaylist(let id, _, _), .removeFromPlaylist(let id, _),
                .moveInPlaylist(let id, _, _), .reorderPlaylist(let id, _, _):
                return id == playlistId
            default:
                return false
            }
        }
    }

    /// Whether an Unsubscribe for `podcastId` is still queued — the podcasts
    /// tombstone set keeps its entries until their op drains (web: the
    /// worker's `unsubscribed` set outlives the op the same way).
    func hasPendingUnsubscribe(podcastId: Int32) -> Bool {
        ops.contains {
            if case .unsubscribe(let id) = $0.kind { return id == podcastId }
            return false
        }
    }

    /// Whether any queued op still reorders the playlists themselves — the
    /// playlists-list refresh keeps the local (optimistic) order until the
    /// moves drain, so server truth can't snap a drag back.
    var hasPendingPlaylistMoves: Bool {
        ops.contains {
            if case .movePlaylist = $0.kind { return true }
            return false
        }
    }

    func clearAll() async {
        await gate.acquire()
        defer { Task { await gate.release() } }
        do {
            try await queue.clearAll()
            try await store.saveDurably(Optional<String>.none, key: "sync-projected-cursor")
            ops.removeAll()
            try await store.saveDurably(ops, key: Self.key)
        } catch { DeviceLog.error("sync clear failed: \(error)") }
    }

    @discardableResult
    func enqueue(_ kind: OutboxOp.Kind) async -> Bool { await enqueueBatch([kind]) }

    func enqueueBatch(_ kinds: [OutboxOp.Kind]) async -> Bool {
        await gate.acquire()
        defer { Task { await gate.release() } }
        guard !Task.isCancelled else { return false }
        let operations = kinds.map(OutboxOp.init)
        let previousIds = Set(ops.map { $0.id.uuidString })
        do {
            try await queue.importOperations(operations: operations.map { try $0.sharedOperation() })
            try await store.projectJournal(operations)
            let pending = Set(try await queue.pendingIds())
            ops.removeAll { previousIds.contains($0.id.uuidString) && !pending.contains($0.id.uuidString) }
            ops.append(contentsOf: operations.filter { pending.contains($0.id.uuidString) })
            // This is a derived mirror; a missing write is rebuilt from the journal on launch.
            await store.save(ops, key: Self.key)
        } catch {
            DeviceLog.error("sync persistence failed: \(error)")
            await MainActor.run { ToastCenter.shared.error("Couldn't save this change. Please try again.") }
            return false
        }
        Task {
            await Task.yield(); await drain()
        }
        return true
    }

    var pendingOperations: [OutboxOp] { ops }

    func setSuspended(_ value: Bool) { suspended = value }

    func drain() async {
        guard !suspended, !draining else { return }
        draining = true
        await gate.acquire()
        defer { draining = false; Task { await gate.release() } }
        guard !suspended else { return }
        do {
            let submitted = Set(ops.map { $0.id.uuidString })
            try await queue.importOperations(operations: ops.map { try $0.sharedOperation() })
            let report = try await perform(queue)
            let recovery = try OutboxOp.restoreSharedQueue(await queue.pending())
            try await store.projectJournal(recovery)
            let projected = Set(recovery.map { $0.id.uuidString })
            let delivered = try await queue.deliveredIds().filter { projected.contains($0) }
            try await queue.confirmCached(sourceIds: delivered)
            try await store.forgetJournalMarkers(Set(delivered))
            var snapshotChanged = false
            if !report.authPaused, !suspended {
                snapshotChanged = try await store.projectSnapshot(
                    synchronize(queue, store.load(String.self, key: "sync-projected-cursor")))
            }
            let pending = Set(try await queue.pendingIds())
            let rejected = Set(try await queue.quarantinedIds())
            let failed = ops.filter { submitted.contains($0.id.uuidString) && rejected.contains($0.id.uuidString) }
            ops.removeAll { submitted.contains($0.id.uuidString) && !pending.contains($0.id.uuidString) }
            try await store.saveDurably(ops, key: Self.key)
            if snapshotChanged { await onSnapshot() }
            for op in failed {
                await onDeadLetter(
                    op,
                    NSError(
                        domain: "HalogenSync", code: 1,
                        userInfo: [NSLocalizedDescriptionKey: report.lastError ?? "Change rejected by the server"]))
            }
        } catch { DeviceLog.warn("sync paused: \(error)") }
    }
}
